### ISSUE TITLE
UPF in Abinit

### ISSUE THREAD

**Issue by gulans** [abinit/abinit#36]
Abinit is able to read UPF pseudopotentials prepared according to the old standard, but it does not support the current standard (version 2.0.1). If I download pseudopotentials in this format from http://www.pseudo-dojo.org/ and try to use it with Abinit, a calculation would crash with the following message:

 Please give name of formatted atomic psp file (and finish with a newline character)
 For atom type 1, psp file is /home/andris/abinit/mpc/psp/O.upf

--- !ERROR
src_file: m_pspheads.F90
src_line: 522
mpi_rank: 0
message: |
    Bad real number in item 1 of list input
...

**Comment by gmatteo** [comment]
The UPF2 format is not supported by Abinit, only UPF1.

At present, it's not clear whether it makes sense to work on the integration of UPF2 in Abinit as I've been told that the QE developers  are working on a new set of the specifications for their pseudopotentials.
Moreover, QE uses a different approach for dealing with relativistic pseudos so even if we start to support UPF2 for scalar pseudos, I don't think we can provide full compatibility for relativistic pseudos. 

The PseudoDojo project uses a very pragmatic approach and provides the same pseudopotential in different formats:

- `psp8` (Abinit-specific format, some optimisations in Abinit are only available if this format is used)
- `psml` (this format is supported by Siesta and Abinit. Note that Abinit must be linked with the optional psml library)
- `upf2` format (QE and other codes compatible with this format)

I'd love to be able to use the same pseudopotential file with different codes but , at present, it seems there's no
agreement on the specifications also because different codes implement things in a different way (the same problem is present for PAW).
In principle, psml can be used with Abinit and QE but, as far as I know, QE does not support psml.

In a nutshell, I would suggest to use either psp8 or psml with Abinit

**Comment by gulans** [comment]
Matteo, thanks for the explanation. If the Abinit developer community has a consensus not to support UPF2 in the foreseeable future, it is perfectly fine. But it would be useful for Abinit users, if either (i) it were documented somewhere or, preferably, (ii) Abinit would recognize UPF2 and refuse to proceed explaining the reason. It took me a while to figure out what is the status with UPFs in Abinit. There are sources on the web like the ESL page stating that UPF is supported in Abinit. On the other hand, pages in the discussion forum explain that Abinit does not support fully-relativistic pseudopotentials specified in UPF. As I had just a scalar-relativistic pseudopotential, I couldn't figure out what is happening until I started "debugging".

**Comment by gmatteo** [comment]
Hi Andris,

> (i) it were documented somewhere

I've documented the formats supported by Abinit  in [this page](https://docs.abinit.org/topics/PseudosPAW/)
of the documentation. The changes will be available when we deploy the new version of the website.

> Abinit would recognize UPF2 and refuse to proceed explaining the reason.

In v9.2, the code will stop with an error message if UPF2 is detected. 

> There are sources on the web like the ESL page stating that UPF is supported in Abinit.

I think you refer to [this page](https://esl.cecam.org/Unified_Pseudopotential_Format_(UPF)).

In this case, one should clarify that only UPF1 is supported by Abinit.
Moreover the statement that UPF is supported by Abinit and GPAW is too vague. 
The UPF format can deal with NC, Ultrasoft and PAW but this does not mean that a code that can read UPF1 or UPF2 files is also able to run calculations with that particular formalism.  I don't think that GPAW can use a NC/ultrasoft pseudo in UPF2 format.
At the same time, Abinit is not able to use PAW or ultrasoft pseudos even if they are provided in UPF1 format.

**Comment by gulans** [comment]
> In v9.2, the code will stop with an error message if UPF2 is detected.
Does it mean we can close the issue?

> I think you refer to this page.
Yes, this is exactly the one I had in mind.

**Comment by gulans** [comment]
> In v9.2, the code will stop with an error message if UPF2 is detected.

Does it mean we can close the issue?

> I think you refer to this page.

Yes, this is exactly the one I had in mind.
